keps/sig-apps/4563-evacuation-api/README.md

```diff
@@ -0,0 +1,1561 @@
+
+There can be multiple evacuators for a single pod, which can be given control of the evacuation by
+the evacuation controller.
+
+We can distinguish 4 different kinds of evacuators:
+- `partial`: it handles only a part of the evacuation process (e.g. releasing one kind of resource).
+  This evacuator should not terminate the pod.
+- `knowledgeable`: has more knowledge (e.g. specific to the deployed application) about how to
+  handle a more graceful evacuation than the controller of the pod (e.g. StatefulSet, Deployment).
+  This can result in the termination of the pod.
+- `controller`: handles full graceful evacuation in a generic way and should result in a pod
+  termination. It can be preempted by a `knowledgeable` evacuator.
+- `fallback`: can handle evacuation if all previous evacuators have failed. This means that
+  `knowledgeable` or `controller` evacuators have either reported no progress, or are simply not
+  implemented for the target pod.
+
+The evacuator should first mark the pods it is interested in evacuating (either partially or fully)
+with the `evacuation.coordination.k8s.io/priority_${EVACUATOR_CLASS}: ${PRIORITY}/${ROLE}`
+annotation. This annotation is parsed into the `Evacuator` type in the [Evacuation API](#evacuation-api).
+
+- `EVACUATOR_CLASS`: should be unique evacuator's DNS subdomain, the maximum length is 54
+  characters (`63 - len("priority_")`)
+- `PRIORITY` and `ROLE`
+  - `controller` should always set a `PRIORITY=10000` and `ROLE=controller`
+  - other evacuators should set `PRIORITY` according to their own needs (minimum value is 0,
+    maximum value is 100000). They can use the `controller` evacuator as a reference point, if
+    they want to be run before or after the `controller` evacuator. They can also observe pod
+    annotations and detect what other evacuators have been registered for the evacuation process.
+    `ROLE` is optional and can be used as a signal to other evacuators. The `controller` value is
+    reserved for pod controllers, but otherwise there is no guidance on how the third party
+    evacuators should name their role.
+
+Example pod with evacuator annotations:
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  annotations:
+    evacuation.coordination.k8s.io/priority_fallback-evacuator.rescue-company.com: "2000"
+    evacuation.coordination.k8s.io/priority_deployment.apps.k8s.io: "10000/controller"
+    evacuation.coordination.k8s.io/priority_sensitive-workload-operator.fruit-company.com: "11000/knowledgeable-app-specific"
+  labels:
+    app: nginx
+  name: sensitive-app
+  namespace: blueberry
+```
+
+The evacuator should observe the evacuation objects that match the pods that the evacuator manages.
+It should start the evacuation only if it observes `.status.activeEvacuatorClass` that matches the
+`EVACUATOR_CLASS` it previously set in the annotation.
+
+If the evacuator is not interested in evacuating the pod, it should set
```

Comment:
Elsewhere you talked about an evacuation not being necessary and being cancelled.  Can it be cancelled once evacuators have started running?  What if they have done something irreversible?  Are they all expected to be reversible?  Idempotent?  Do they get called to reverse an evacuation?

Reply:
The inspiration for this was (partly) the declarative node maintenance KEP - https://github.com/kubernetes/enhancements/issues/4212

With declaring a node as under maintenance it feels implicit that you can end maintenance at any time, either as a cancellation or by marking it complete. That's how organizations getting humans to maintain things typically model it.

(also, you can't “reverse“ a maintenance notification; you can rescind it, but you can't create a second maintenance period that overlaps with and cancels out the first one).

Reply:
My point was that I can't tell, from this KEP, how to answer those questions.  If those answers are in there, they are buried.

Reply:
I have added 3 examples to outline how the EvictionRequest cancellation should work: https://github.com/atiratree/kube-enhancements/blob/evacuation-api/keps/sig-apps/4563-eviction-request-api/README.md#evictionrequest-cancellation-examples